README.md
# Koa Static Cache

[![NPM version][npm-image]][npm-url]
[![build status][travis-image]][travis-url]
[![Test coverage][coveralls-image]][coveralls-url]
[![David deps][david-image]][david-url]

[npm-image]: https://img.shields.io/npm/v/koa-static-cache.svg?style=flat-square
[npm-url]: https://npmjs.org/package/koa-static-cache
[travis-image]: https://img.shields.io/travis/koajs/static-cache.svg?style=flat-square
[travis-url]: https://travis-ci.org/koajs/static-cache
[coveralls-image]: https://img.shields.io/coveralls/koajs/static-cache.svg?style=flat-square
[coveralls-url]: https://coveralls.io/r/koajs/static-cache?branch=master
[david-image]: https://img.shields.io/david/koajs/static-cache.svg?style=flat-square
[david-url]: https://david-dm.org/koajs/static-cache

Static server for koa.

Differences between this library and other libraries such as [static](https://github.com/koajs/static):

- There is no directory or `index.html` support.
- You may optionally store the data in memory - it streams by default.
- Caches the assets on initialization - you need to restart the process to update the assets.(can turn off with options.preload = false)
- Uses MD5 hash sum as an ETag.
- Uses .gz files if present on disk, like nginx gzip_static module

## Installation

```js
$ npm install koa-static-cache
```

## API

### staticCache(dir [, options] [, files])

```js
var path = require('path')
var staticCache = require('koa-static-cache')

app.use(staticCache(path.join(__dirname, 'public'), {
  maxAge: 365 * 24 * 60 * 60
}))
```

- `dir` (str) - the directory you wish to serve, priority than `options.dir`.
- `options.dir` (str) - the directory you wish to serve, default to `process.cwd`.
- `options.maxAge` (int) - cache control max age for the files, `0` by default.
- `options.cacheControl` (str) - optional cache control header. Overrides `options.maxAge`.
- `options.buffer` (bool) - store the files in memory instead of streaming from the filesystem on each request.
- `options.gzip` (bool) - when request's accept-encoding include gzip, files will compressed by gzip.
- `options.usePrecompiledGzip` (bool) - try use gzip files, loaded from disk, like nginx gzip_static
- `options.alias` (obj) - object map of aliases. See below.
- `options.prefix` (str) - the url prefix you wish to add, default to `''`.
- `options.dynamic` (bool) - dynamic load file which not cached on initialization.
- `options.filter` (function | array) - filter files at init dir, for example - skip non build (source) files. If array set - allow only listed files
- `options.preload` (bool) - caches the assets on initialization or not, default to `true`. always work together with `options.dynamic`.
- `options.files` (obj) - optional files object. See below.
- `files` (obj) - optional files object. See below.

### Aliases

For example, if you have this alias object:

```js
{
  '/favicon.png': '/favicon-32.png'
}
```

Then requests to `/favicon.png` will actually return `/favicon-32.png` without redirects or anything.
This is particularly important when serving [favicons](https://github.com/audreyr/favicon-cheat-sheet) as you don't want to store duplicate images.

### Files

You can pass in an optional files object.
This allows you to do two things:

#### Combining directories into a single middleware

Instead of doing:

```js
app.use(staticCache('/public/js'))
app.use(staticCache('/public/css'))
```

You can do this:

```js
var files = {}

// Mount the middleware
app.use(staticCache('/public/js', {}, files))

// Add additional files
staticCache('/public/css', {}, files)
```

The benefit is that you'll have one less function added to the stack as well as doing one hash lookup instead of two.

#### Editing the files object

For example, if you want to change the max age of `/package.json`, you can do the following:

```js
var files = {}

app.use(staticCache('/public', {
  maxAge: 60 * 60 * 24 * 365
}, files))

files['/package.json'].maxAge = 60 * 60 * 24 * 30
```

#### Using a LRU cache to avoid OOM when dynamic mode enabled

You can pass in a lru cache instance which has tow methods: `get(key)` and `set(key, value)`.

```js
var LRU = require('lru-cache')
var files = new LRU({ max: 1000 })

app.use(staticCache({
  dir: '/public',
  dynamic: true,
  files: files
}))
```

## License

The MIT License (MIT)

Copyright (c) 2013 Jonathan Ong me@jongleberry.com

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.


HISTORY.md

5.1.4 / 2020-08-03
==================

**fixes**
  * [[`2735fab`](http://github.com/koajs/static-cache/commit/2735fab6b9303da8ea941eab66f719440e0f2763)] - fix: remove unused require (dead-horse <<dead_horse@qq.com>>)
  * [[`e12d920`](http://github.com/koajs/static-cache/commit/e12d9202e4900f17f7808e505ae39c161615be7f)] - fix: correct the time compare (dead-horse <<dead_horse@qq.com>>)
  * [[`c24e6c3`](http://github.com/koajs/static-cache/commit/c24e6c3fe59b896b92a69045deaa0766f6c40836)] - fix: mtime (#93) (Jlin <<565774991@qq.com>>)

5.1.3 / 2020-04-29
==================

**fixes**
  * [[`dca1edf`](http://github.com/koajs/static-cache/commit/dca1edf0b641993921b33c0289dbd73ff4bc1c13)] - fix: alias should work when preload = false (#84) (TZ | 天猪 <<atian25@qq.com>>)

5.1.2 / 2018-02-06
==================

**others**
  * [[`ba60f3d`](http://github.com/koajs/static-cache/commit/ba60f3d859e98efd41a625fd0410ff4d930bf37b)] - deps: use ^ (dead-horse <<dead_horse@qq.com>>)
  * [[`82274d0`](http://github.com/koajs/static-cache/commit/82274d02d3601282cb363a2339081393ef2bf83d)] - deps: Update package.json bump debug and mz versions (#73) (Iain Maitland <<imaitland@protonmail.com>>)

5.1.1 / 2017-06-13
==================

  * fix: only load file under options.dir (#67)

5.1.0 / 2017-06-01
==================

  * feat: files support lru (#64)
  * Update mz (#61)

5.0.1 / 2017-04-19
==================

  * Add node.js v7.6.0 support (#58)

5.0.0 / 2017-04-01
==================

  * Support Koa 2 (#57)

4.0.0 / 2017-02-21
==================

  * refactor: check prefix first to avoid calculate (#56)

3.2.0 / 2017-01-07
==================

  * feat: support options.preload (#55)

3.1.7 / 2016-04-07
==================

  * update mz to 2.4.0

3.1.6 / 2016-03-22
==================

  * fix: don't catch downstream error

3.1.5 / 2016-03-02
==================

  * fix: dynamic load file bug on windows platform

3.1.4 / 2016-01-04
==================

  * bump deps
  * use mz

3.1.3 / 2015-11-26
==================

  * Fix broken mtime comparison

3.1.2 / 2015-07-08
==================

  * bugfix: error on dynamic files

3.1.1 / 2015-04-17
==================

  * fix: options.prefix bug in windows, fixes #39

3.1.0 / 2015-03-28
==================

  * Merge pull request #33 from AlexeyKhristov/gz

3.0.3 / 2015-03-28
==================

  * fix problem, cache is not used in dynamic mode

3.0.2 / 2015-03-18
==================

  * fix options.prefix bug in windows, fixes #36

3.0.1 / 2015-01-06
==================

  * feat(dynamic): add dynamic option to support dynamic load
  * fix(dynamic): use stat to detect folder

3.0.0 / 2015-01-06
==================

  * fix(test): typo
  * fix(buffer): keep the old logic of treat unbuffered file
  * feat: add opt.buffer false to serve file not cache at all
  * fix: support load file dynamic, close #30

2.0.2 / 2015-01-05
==================

  * fix normalize bug in windows, fixes #29

2.0.1 / 2014-12-02
==================

  * accept abnormal path, like: //index.html

2.0.0 / 2014-11-14
==================

  * bump koa
  * only response GET and HEAD

1.2.0 / 2014-09-18
==================

  * bump compressible and mime-types
  * decodeURI when use this.path as key to fetch value from files object

1.1.0 / 2014-07-16
==================

  * replace mime by mime-types
  * remove onerror and destroy, let koa hanlde these stuff

1.0.10 / 2014-05-18
==================

  * bump fs-readdir-recursive, fixed #14
  * fix bad argument handling, fixed #20
  * should not return gzip buffer when accept encoding not include gzip

1.0.9 / 2014-03-31
==================

  * add url prefix option

1.0.8 / 2014-03-31
==================

  * support options.dir, default to process.cwd()
  * add vary, check file's length when gzip
  * Ensure files can be gzipped via compressible.

1.0.7 / 2014-03-26
==================

  * add options.gzip to control gzip, support stream's gzip
  * add gzip support for buffers

1.0.3 / 2014-01-14
==================

 * update `on-socket-error`

1.0.0 / 2013-12-21
==================

 * use `yield* next`


test/index.js
var fs = require('fs')
var crypto = require('crypto')
var zlib = require('zlib')
var request = require('supertest')
var should = require('should')
var Koa = require('koa')
var http = require('http')
var path = require('path')
var staticCache = require('..')
var LRU = require('ylru')

var app = new Koa()
var files = {}
app.use(staticCache(path.join(__dirname, '..'), {
  alias: {
    '/package': '/package.json'
  },
  filter(file) {
    return !file.includes('node_modules')
  }
}, files))

var server = http.createServer(app.callback())

var app2 = new Koa()
app2.use(staticCache(path.join(__dirname, '..'), {
  buffer: true,
  filter(file) {
    return !file.includes('node_modules')
  }
}))
var server2 = http.createServer(app2.callback())

var app3 = new Koa()
app3.use(staticCache(path.join(__dirname, '..'), {
  buffer: true,
  gzip: true,
  filter(file) {
    return !file.includes('node_modules')
  }
}))
var server3 = http.createServer(app3.callback())

var app4 = new Koa()
var files4 = {}
app4.use(staticCache(path.join(__dirname, '..'), {
  gzip: true,
  filter(file) {
    return !file.includes('node_modules')
  }
}, files4))

var server4 = http.createServer(app4.callback())

var app5 = new Koa()
app5.use(staticCache({
  buffer: true,
  prefix: '/static',
  dir: path.join(__dirname, '..'),
  filter(file) {
    return !file.includes('node_modules')
  }
}))
var server5 = http.createServer(app5.callback())

describe('Static Cache', function () {

  it('should dir priority than options.dir', function (done) {
    var app = new Koa()
    app.use(staticCache(path.join(__dirname, '..'), {
      dir: __dirname
    }))
    var server = app.listen()
    request(server)
    .get('/index.js')
    .expect(200, done)
  })

  it('should default options.dir works fine', function (done) {
    var app = new Koa()
    app.use(staticCache({
      dir: path.join(__dirname, '..')
    }))
    var server = app.listen()
    request(server)
    .get('/index.js')
    .expect(200, done)
  })

  it('should accept abnormal path', function (done) {
    var app = new Koa()
    app.use(staticCache({
      dir: path.join(__dirname, '..')
    }))
    var server = app.listen()
    request(server)
    .get('//index.js')
    .expect(200, done)
  })

  it('should default process.cwd() works fine', function (done) {
    var app = new Koa()
    app.use(staticCache())
    var server = app.listen()
    request(server)
    .get('/index.js')
    .expect(200, done)
  })

  var etag
  it('should serve files', function (done) {
    request(server)
    .get('/index.js')
    .expect(200)
    .expect('Cache-Control', 'public, max-age=0')
    .expect('Content-Type', /javascript/)
    .end(function (err, res) {
      if (err)
        return done(err)

      res.should.have.header('Content-Length')
      res.should.have.header('Last-Modified')
      res.should.have.header('ETag')
      etag = res.headers.etag

      done()
    })
  })

  it('should serve files as buffers', function (done) {
    request(server2)
    .get('/index.js')
    .expect(200)
    .expect('Cache-Control', 'public, max-age=0')
    .expect('Content-Type', /javascript/)
    .end(function (err, res) {
      if (err)
        return done(err)

      res.should.have.header('Content-Length')
      res.should.have.header('Last-Modified')
      res.should.have.header('ETag')

      etag = res.headers.etag

      done()
    })
  })

  it('should serve recursive files', function (done) {
    request(server)
    .get('/test/index.js')
    .expect(200)
    .expect('Cache-Control', 'public, max-age=0')
    .expect('Content-Type', /javascript/)
    .end(function (err, res) {
      if (err)
        return done(err)

      res.should.have.header('Content-Length')
      res.should.have.header('Last-Modified')
      res.should.have.header('ETag')

      done()
    })
  })

  it('should not serve hidden files', function (done) {
    request(server)
    .get('/.gitignore')
    .expect(404, done)
  })

  it('should support conditional HEAD requests', function (done) {
    request(server)
    .head('/index.js')
    .set('If-None-Match', etag)
    .expect(304, done)
  })

  it('should support conditional GET requests', function (done) {
    request(server)
    .get('/index.js')
    .set('If-None-Match', etag)
    .expect(304, done)
  })

  it('should support HEAD', function (done) {
    request(server)
    .head('/index.js')
    .expect(200)
    .expect('', done)
  })

  it('should support 404 Not Found for other Methods to allow downstream',
  function (done) {
    request(server)
    .put('/index.js')
    .expect(404, done)
  })

  it('should ignore query strings', function (done) {
    request(server)
    .get('/index.js?query=string')
    .expect(200, done)
  })

  it('should alias paths', function (done) {
    request(server)
    .get('/package')
    .expect('Content-Type', /json/)
    .expect(200, done)
  })

  it('should be configurable via object', function (done) {
    files['/package.json'].maxAge = 1

    request(server)
    .get('/package.json')
    .expect('Cache-Control', 'public, max-age=1')
    .expect(200, done)
  })

  it('should set the etag header', function (done) {
    var pk = fs.readFileSync('package.json')
    var md5 = crypto.createHash('md5').update(pk).digest('base64')

    request(server)
    .get('/package.json')
    .expect('ETag', '"' + md5 + '"')
    .expect(200, done)
  })

  it('should set Last-Modified if file modified and not buffered', function (done) {
    setTimeout(function () {
      var readme = fs.readFileSync('README.md', 'utf8')
      fs.writeFileSync('README.md', readme, 'utf8')
      var mtime = fs.statSync('README.md').mtime
      var md5 = files['/README.md'].md5
      request(server)
      .get('/README.md')
      .expect(200, function (err, res) {
        res.should.have.header('Content-Length')
        res.should.have.header('Last-Modified')
        res.should.not.have.header('ETag')
        files['/README.md'].mtime.should.eql(mtime)
        setTimeout(function () {
          files['/README.md'].md5.should.equal(md5)
        }, 10)
        done()
      })
    }, 1000)
  })

  it('should set Last-Modified if file rollback and not buffered', function (done) {
    setTimeout(function () {
      var readme = fs.readFileSync('README.md', 'utf8')
      fs.writeFileSync('README.md', readme, 'utf8')
      var mtime = fs.statSync('README.md').mtime
      var md5 = files['/README.md'].md5
      request(server)
      .get('/README.md')
      .expect(200, function (err, res) {
        res.should.have.header('Content-Length')
        res.should.have.header('Last-Modified')
        res.should.not.have.header('ETag')
        files['/README.md'].mtime.should.eql(mtime)
        setTimeout(function () {
          files['/README.md'].md5.should.equal(md5)
        }, 10)
        done()
      })
    }, 1000)
  })

  it('should serve files with gzip buffer', function (done) {
    var index = fs.readFileSync('index.js')
    zlib.gzip(index, function (err, content) {
      request(server3)
      .get('/index.js')
      .set('Accept-Encoding', 'gzip')
      .expect(200)
      .expect('Cache-Control', 'public, max-age=0')
      .expect('Content-Encoding', 'gzip')
      .expect('Content-Type', /javascript/)
      .expect('Content-Length', content.length)
      .expect('Vary', 'Accept-Encoding')
      .expect(index.toString())
      .end(function (err, res) {
        if (err)
          return done(err)
        res.should.have.header('Content-Length')
        res.should.have.header('Last-Modified')
        res.should.have.header('ETag')

        etag = res.headers.etag

        done()
      })
    })
  })

  it('should not serve files with gzip buffer when accept encoding not include gzip',
  function (done) {
    var index = fs.readFileSync('index.js')
    request(server3)
    .get('/index.js')
    .set('Accept-Encoding', '')
    .expect(200)
    .expect('Cache-Control', 'public, max-age=0')
    .expect('Content-Type', /javascript/)
    .expect('Content-Length', index.length)
    .expect('Vary', 'Accept-Encoding')
    .expect(index.toString())
    .end(function (err, res) {
      if (err)
        return done(err)
      res.should.not.have.header('Content-Encoding')
      res.should.have.header('Content-Length')
      res.should.have.header('Last-Modified')
      res.should.have.header('ETag')
      done()
    })
  })

  it('should serve files with gzip stream', function (done) {
    var index = fs.readFileSync('index.js')
    zlib.gzip(index, function (err, content) {
      request(server4)
      .get('/index.js')
      .set('Accept-Encoding', 'gzip')
      .expect(200)
      .expect('Cache-Control', 'public, max-age=0')
      .expect('Content-Encoding', 'gzip')
      .expect('Content-Type', /javascript/)
      .expect('Vary', 'Accept-Encoding')
      .expect(index.toString())
      .end(function (err, res) {
        if (err)
          return done(err)
        res.should.not.have.header('Content-Length')
        res.should.have.header('Last-Modified')
        res.should.have.header('ETag')

        etag = res.headers.etag

        done()
      })
    })
  })

  it('should serve files with prefix', function (done) {
    request(server5)
    .get('/static/index.js')
    .expect(200)
    .expect('Cache-Control', 'public, max-age=0')
    .expect('Content-Type', /javascript/)
    .end(function (err, res) {
      if (err)
        return done(err)

      res.should.have.header('Content-Length')
      res.should.have.header('Last-Modified')
      res.should.have.header('ETag')

      etag = res.headers.etag

      done()
    })
  })

  it('should 404 when dynamic = false', function (done) {
    var app = new Koa()
    app.use(staticCache({dynamic: false}))
    var server = app.listen()
    fs.writeFileSync('a.js', 'hello world')

    request(server)
      .get('/a.js')
      .expect(404, function(err) {
        fs.unlinkSync('a.js')
        done(err)
      })
  })

  it('should work fine when new file added in dynamic mode', function (done) {
    var app = new Koa()
    app.use(staticCache({dynamic: true}))
    var server = app.listen()
    fs.writeFileSync('a.js', 'hello world')

    request(server)
      .get('/a.js')
      .expect(200, function(err) {
        fs.unlinkSync('a.js')
        done(err)
      })
  })

  it('should work fine when new file added in dynamic and prefix mode', function (done) {
    var app = new Koa()
    app.use(staticCache({dynamic: true, prefix: '/static'}))
    var server = app.listen()
    fs.writeFileSync('a.js', 'hello world')

    request(server)
      .get('/static/a.js')
      .expect(200, function(err) {
        fs.unlinkSync('a.js')
        done(err)
      })
  })

  it('should work fine when new file added in dynamic mode with LRU', function (done) {
    var app = new Koa()
    var files = new LRU(1)
    app.use(staticCache({dynamic: true, files: files}))
    var server = app.listen()
    fs.writeFileSync('a.js', 'hello world a')
    fs.writeFileSync('b.js', 'hello world b')
    fs.writeFileSync('c.js', 'hello world b')

    request(server)
      .get('/a.js')
      .expect(200, function(err) {
        should.exist(files.get('/a.js'))
        should.not.exist(err)

        request(server)
          .get('/b.js')
          .expect(200, function (err) {
            should.not.exist(files.get('/a.js'))
            should.exist(files.get('/b.js'))
            should.not.exist(err)

            request(server)
              .get('/c.js')
              .expect(200, function (err) {
                should.not.exist(files.get('/b.js'))
                should.exist(files.get('/c.js'))
                should.not.exist(err)

                request(server)
                  .get('/a.js')
                  .expect(200, function (err) {
                    should.not.exist(files.get('/c.js'))
                    should.exist(files.get('/a.js'))
                    should.not.exist(err)
                    fs.unlinkSync('a.js')
                    fs.unlinkSync('b.js')
                    fs.unlinkSync('c.js')
                    done()
                  })
              })
          })
      })
  })

  it('should 404 when url without prefix in dynamic and prefix mode', function (done) {
    var app = new Koa()
    app.use(staticCache({dynamic: true, prefix: '/static'}))
    var server = app.listen()
    fs.writeFileSync('a.js', 'hello world')

    request(server)
      .get('/a.js')
      .expect(404, function(err) {
        fs.unlinkSync('a.js')
        done(err)
      })
  })

  it('should 404 when new hidden file added in dynamic mode', function (done) {
    var app = new Koa()
    app.use(staticCache({dynamic: true}))
    var server = app.listen()
    fs.writeFileSync('.a.js', 'hello world')

    request(server)
      .get('/.a.js')
      .expect(404, function(err) {
        fs.unlinkSync('.a.js')
        done(err)
      })
  })

  it('should 404 when file not exist in dynamic mode', function (done) {
    var app = new Koa()
    app.use(staticCache({dynamic: true}))
    var server = app.listen()
    request(server)
      .get('/a.js')
      .expect(404, done)
  })

  it('should 404 when file not exist', function (done) {
    var app = new Koa()
    app.use(staticCache({dynamic: true}))
    var server = app.listen()
    request(server)
      .get('/a.js')
      .expect(404, done)
  })

  it('should 404 when is folder in dynamic mode', function (done) {
    var app = new Koa()
    app.use(staticCache({dynamic: true}))
    var server = app.listen()
    request(server)
      .get('/test')
      .expect(404, done)
  })

  it('should array options.filter works fine', function (done) {
    var app = new Koa()
    app.use(staticCache({
      dir: path.join(__dirname, '..'),
      filter: ['index.js']
    }))
    var server = app.listen()
    request(server)
    .get('/Makefile')
    .expect(404, done)
  })

  it('should function options.filter works fine', function (done) {
    var app = new Koa()
    app.use(staticCache({
      dir: path.join(__dirname, '..'),
      filter: function (file) { return file.indexOf('index.js') === 0 }
    }))
    var server = app.listen()
    request(server)
    .get('/Makefile')
    .expect(404, done)
  })

  it('should options.dynamic and options.preload works fine', function (done) {
    var app = new Koa()
    var files = {}
    app.use(staticCache({
      dir: path.join(__dirname, '..'),
      preload: false,
      dynamic: true,
      files: files
    }))
    files.should.eql({})
    request(app.listen())
      .get('/Makefile')
      .expect(200, function (err, res) {
        should.not.exist(err)
        files.should.have.keys('/Makefile')
        done()
      })
  })

  it('should options.alias and options.preload works fine', function (done) {
    var app = new Koa()
    var files = {}
    app.use(staticCache({
      dir: path.join(__dirname, '..'),
      preload: false,
      dynamic: true,
      alias: {
        '/package': '/package.json'
      },
      files: files
    }))
    files.should.eql({})
    request(app.listen())
      .get('/package')
      .expect(200, function (err, res) {
        if (err) return done(err)
        should.not.exist(err)
        files.should.have.keys('/package.json')
        files.should.not.have.keys('/package')

        request(app.listen())
          .get('/package.json')
          .expect(200, function (err, res) {
            if (err) return done(err)
            should.not.exist(err)
            files.should.have.keys('/package.json')
            should.ok(Object.keys(files).length === 1)
            done()
          })
      })
  })


  it('should loadFile under options.dir', function (done) {
    var app = new Koa()
    app.use(staticCache({
      dir: __dirname,
      preload: false,
      dynamic: true,
    }))
    request(app.listen())
      .get('/%2E%2E/package.json')
      .expect(404)
      .end(done)
  })
})


index.js
var crypto = require('crypto')
var fs = require('mz/fs')
var zlib = require('mz/zlib')
var path = require('path')
var mime = require('mime-types')
var compressible = require('compressible')
var readDir = require('fs-readdir-recursive')
var debug = require('debug')('koa-static-cache')

module.exports = function staticCache(dir, options, files) {
  if (typeof dir === 'object') {
    files = options
    options = dir
    dir = null
  }

  options = options || {}
  // prefix must be ASCII code
  options.prefix = (options.prefix || '').replace(/\/*$/, '/')
  files = new FileManager(files || options.files)
  dir = dir || options.dir || process.cwd()
  dir = path.normalize(dir)
  var enableGzip = !!options.gzip
  var filePrefix = path.normalize(options.prefix.replace(/^\//, ''))

  // option.filter
  var fileFilter = function () { return true }
  if (Array.isArray(options.filter)) fileFilter = function (file) { return ~options.filter.indexOf(file) }
  if (typeof options.filter === 'function') fileFilter = options.filter

  if (options.preload !== false) {
    readDir(dir).filter(fileFilter).forEach(function (name) {
      loadFile(name, dir, options, files)
    })
  }

  return async (ctx, next) => {
    // only accept HEAD and GET
    if (ctx.method !== 'HEAD' && ctx.method !== 'GET') return await next()
    // check prefix first to avoid calculate
    if (ctx.path.indexOf(options.prefix) !== 0) return await next()

    // decode for `/%E4%B8%AD%E6%96%87`
    // normalize for `//index`
    var filename = path.normalize(safeDecodeURIComponent(ctx.path))

    // check alias
    if (options.alias && options.alias[filename]) filename = options.alias[filename];

    var file = files.get(filename)
    // try to load file
    if (!file) {
      if (!options.dynamic) return await next()
      if (path.basename(filename)[0] === '.') return await next()
      if (filename.charAt(0) === path.sep) filename = filename.slice(1)

      // trim prefix
      if (options.prefix !== '/') {
        if (filename.indexOf(filePrefix) !== 0) return await next()
        filename = filename.slice(filePrefix.length)
      }

      var fullpath = path.join(dir, filename)
      // files that can be accessd should be under options.dir
      if (fullpath.indexOf(dir) !== 0) {
        return await next()
      }

      var s
      try {
        s = await fs.stat(fullpath)
      } catch (err) {
        return await next()
      }
      if (!s.isFile()) return await next()

      file = loadFile(filename, dir, options, files)
    }

    ctx.status = 200

    if (enableGzip) ctx.vary('Accept-Encoding')

    if (!file.buffer) {
      var stats = await fs.stat(file.path)
      if (stats.mtime.getTime() !== file.mtime.getTime()) {
        file.mtime = stats.mtime
        file.md5 = null
        file.length = stats.size
      }
    }

    ctx.response.lastModified = file.mtime
    if (file.md5) ctx.response.etag = file.md5

    if (file.md5 && ctx.fresh) return ctx.status = 304

    ctx.type = file.type
    ctx.length = file.zipBuffer ? file.zipBuffer.length : file.length
    ctx.set('cache-control', file.cacheControl || 'public, max-age=' + file.maxAge)
    if (file.md5) ctx.set('content-md5', file.md5)

    if (ctx.method === 'HEAD') return

    var acceptGzip = ctx.acceptsEncodings('gzip') === 'gzip'

    if (file.zipBuffer) {
      if (acceptGzip) {
        ctx.set('content-encoding', 'gzip')
        ctx.body = file.zipBuffer
      } else {
        ctx.body = file.buffer
      }
      return
    }

    var shouldGzip = enableGzip
      && file.length > 1024
      && acceptGzip
      && compressible(file.type)

    if (file.buffer) {
      if (shouldGzip) {

        var gzFile = files.get(filename + '.gz')
        if (options.usePrecompiledGzip && gzFile && gzFile.buffer) { // if .gz file already read from disk
          file.zipBuffer = gzFile.buffer
        } else {
          file.zipBuffer = await zlib.gzip(file.buffer)
        }
        ctx.set('content-encoding', 'gzip')
        ctx.body = file.zipBuffer
      } else {
        ctx.body = file.buffer
      }
      return
    }

    var stream = fs.createReadStream(file.path)

    // update file hash
    if (!file.md5) {
      var hash = crypto.createHash('md5')
      stream.on('data', hash.update.bind(hash))
      stream.on('end', function () {
        file.md5 = hash.digest('base64')
      })
    }

    ctx.body = stream
    // enable gzip will remove content length
    if (shouldGzip) {
      ctx.remove('content-length')
      ctx.set('content-encoding', 'gzip')
      ctx.body = stream.pipe(zlib.createGzip())
    }
  }
}

function safeDecodeURIComponent(text) {
  try {
    return decodeURIComponent(text)
  } catch (e) {
    return text
  }
}

/**
 * load file and add file content to cache
 *
 * @param {String} name
 * @param {String} dir
 * @param {Object} options
 * @param {Object} files
 * @return {Object}
 * @api private
 */

function loadFile(name, dir, options, files) {
  var pathname = path.normalize(path.join(options.prefix, name))
  if (!files.get(pathname)) files.set(pathname, {})
  var obj = files.get(pathname)
  var filename = obj.path = path.join(dir, name)
  var stats = fs.statSync(filename)
  var buffer = fs.readFileSync(filename)

  obj.cacheControl = options.cacheControl
  obj.maxAge = obj.maxAge ? obj.maxAge : options.maxAge || 0
  obj.type = obj.mime = mime.lookup(pathname) || 'application/octet-stream'
  obj.mtime = stats.mtime
  obj.length = stats.size
  obj.md5 = crypto.createHash('md5').update(buffer).digest('base64')

  debug('file: ' + JSON.stringify(obj, null, 2))
  if (options.buffer)
    obj.buffer = buffer

  buffer = null
  return obj
}

function FileManager(store) {
  if (store && typeof store.set === 'function' && typeof store.get === 'function') {
    this.store = store
  } else {
    this.map = store || Object.create(null)
  }
}

FileManager.prototype.get = function (key) {
  return this.store ? this.store.get(key) : this.map[key]
}

FileManager.prototype.set = function (key, value) {
  if (this.store) return this.store.set(key, value)
  this.map[key] = value
}
